RFC: Add instance type settings design proposal (#2390)
aws/karpenter: designs/instance-type-settings.md

    hardware-vendor.example/foo: "2"
    hardware-vendor.example/bar: "20"
    pods: 10
    memory: 3865Mi 
```

**Extend `v1alpha5/Provisioner` CRD**

In particular, we will extend the `.spec.kubeletConfiguration` with `kubeReserved` and `evictionThreshold` that will be passed down into the bootstrap scripts.

```yaml
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  requirements:
    - key: karpenter.sh/capacity-type
      operator: In
      values: ["spot"]
  limits:
    resources:
      cpu: "1000"
  kubeletConfiguration:
    kubeReserved:
      memory: "1Gi"
      cpu: "300m"
      ephemeral-storage: "1Gi"
    systemReserved:
      memory: "1Gi"
      cpu: "300m"
      ephemeral-storage: "1Gi"
    evictionThreshold:
      hard:
        memoryAvailable: "1Gi"
        imagefsAvailable: "10%"
        nodefsAvailable: "15%"
        nodefsInodesFree: "5%"
      soft:
        ...
```

## Instance Type Setting Examples

1. Configure the VM memory capacity value for `c5.large`

```yaml
apiVersion: karpenter.sh/v1alpha1
kind: InstanceType
metadata:
   name: "c5.large"
spec:
   capacity:
      memory: 200Mi
```

> 1. Configure custom device resources for custom device resources for `c4.large`

Reviewer: ```suggestion
1. Configure custom device resources for `c4.large`
```
I think that's what you mean.

```yaml
apiVersion: karpenter.sh/v1alpha1
kind: InstanceType
metadata:
  name: "c4.large"
spec:
   allocatable:
      hardware-vendor.example/foo: "2"
```

1. Specify custom offerings with custom pricing information for `c4.large`

```yaml
apiVersion: karpenter.sh/v1alpha1
kind: InstanceType
metadata:
  name: "c4.large"
spec:
  offerings:
  - karpenter.sh/capacity-type: on-demand
    topology.kubernetes.io/zone: us-west-2b
    karpenter.sh/price: "0.60"
  - karpenter.sh/capacity-type: spot
    karpenter.sh/price: "0.40"
```

## Considerations

### Instance Type Overrides for Other Cloud Providers (Azure/CAPI/etc.)

> Because Karpenter is intended to be cloud-agnostic, we need to keep this in mind when adding changes to the API surface. For the initial iteration of the `InstanceType` CRD, we will assume that the instance type will be selected by the `.metadata.name` field which will map to the well-defined Kubernetes node field `node.kubernetes.io/instance-type`.

Reviewer 2: ```suggestion
Because Karpenter is cloud-agnostic, we need to keep this in mind when adding changes to the API surface. For the initial iteration of the `InstanceType` CRD, we will assume that the instance type will be selected by the `.metadata.name` field which will map to the well-defined Kubernetes node field `node.kubernetes.io/instance-type`.
```

For other clouds, this is expected to be mapped on the nodes as it is in AWS and the `GetInstanceType` call for other cloud providers is expected to retrieve and return all the instance type values that map to the `node.kubernetes.io/instance-type` value. In this way, Karpenter can perform instance type overrides for other clouds when instance types are surfaced.

### Custom Resource Cardinality

> One of the primary questions presented is whether we should create a 1-1 mapping between settings and a given instance type or whether we should allow the ability to create more complex structures using set relationships and label selectors. Options are listed below combined with the pros and cons for each:

Reviewer: ```suggestion
One of the primary questions presented is whether we should create a one-to-one mapping between settings and a given instance type or whether we should allow the ability to create more complex structures using set relationships and label selectors. Options are listed below, combined with the pros and cons for each:
```

**Options**

> 1. Create an `InstanceType` CRD that specifies a 1-1 relationship between the settings and the instance type

Reviewer: ```suggestion
1. Create an `InstanceType` CRD that specifies a one-to-one relationship between the settings and the instance type
```

    1. Pros
        1. Clearly defined relationship between an instance type and the settings that are defined for it
        2. Users that only have a few instance types will not have an issue configuring the instance types with these values
        3. We can consider adding an `InstanceTypeClass` or similar at a later date that would allow a grouping of instance types to be assigned the same settings
    2. Cons
        1. May not scale well as users that have high levels of customizations across large numbers of instance types may have to maintain a large number of `InstanceType` CRDs

>         2. I may want to define settings over an instance family without having to individually configure the instances that exist within that family

Reviewer: ```suggestion
        2. Someone may want to define settings over an instance family without having to individually configure the instances that exist within that family
```

> 2. Create an `InstanceTypeSetting` CRD that contains a list of instance types that allow a 1-many relationship between settings and instance types. These instance types could be selected either with a `selector` or `regex`

Reviewer: ```suggestion
2. Create an `InstanceTypeSetting` CRD that contains a list of instance types that allow a one-to-many relationship between settings and instance types. These instance types could be selected either with a `selector` or `regex`
```

    1. Pros
        1. Allows users who want to apply the same instance type settings across a wide range of instance types or across an instance type family to do so

>         2. Lower maintenance burden from maintaining less `InstanceTypeSetting` CRDs

Reviewer: ```suggestion
        2. Lower maintenance burden from maintaining fewer `InstanceTypeSetting` CRDs
```

    2. Cons
        1. Can create complex relationships that may be difficult to track for users
        2. If we were to allow overlapping `InstanceTypeSettings`, we would have to create a `weight` on the `InstanceTypeSettings` which would further convolute details for the user

>         3. If we were to deny overlapping `InstanceTypeSettings` , then the feature of specifying a group of instance types may not be used that often as configuration between instance types may rarely be shared

Reviewer: ```suggestion
        3. If we were to deny overlapping `InstanceTypeSettings`, then the feature of specifying a group of instance types may not be used that often, as configuration between instance types may rarely be shared
```

        4. It’s unclear whether users would want have enough common configuration that they would assign across instance types. It makes more sense to iterate on a grouping mechanism later if there is a user ask for it

> **Recommendation:** Use `InstanceType` CRD for initial iteration for 1-1 mappings before considering introducing more complex constructs to the user. While specifying rulesets might reduce the overhead burden for customers to keep track of separate `InstanceTypeSettings` , ultimately without an API surface to view exactly what settings that ruleset will produce, it will ultimately be confusing for users.

Reviewer: ```suggestion
**Recommendation:** Use `InstanceType` CRD for initial iteration for one-to-one mappings before considering introducing more complex constructs to the user. While specifying rulesets might reduce the overhead burden for customers to keep track of separate `InstanceTypeSettings`, ultimately without an API surface to view exactly what settings that ruleset will produce, it will ultimately be confusing for users.
```

> **Recommendation:** Use `InstanceType` CRD for initial iteration for 1-1 mappings before considering introducing more complex constructs to the user. While specifying rulesets might reduce the overhead burden for customers to keep track of separate `InstanceTypeSettings` , ultimately without an API surface to view exactly what settings that ruleset will produce, it will ultimately be confusing for users.

Reviewer 3: For our use case (similar to #1900 where nodes offer fuse devices via the [smarter device manager](https://gitlab.com/arm-research/smarter/smarter-device-manager)) the recommended approach of having a 1-1 mapping would create quite a significant overhead and burden: For every instance type and size, we would need to create an `InstanceType` resource and configure it. Even with templating this is quite a bit of effort and seems prone to errors. Also, we would need to keep the list of available instance types up to date, increasing the maintenance work. So going with the second option sounds like a more sustainable approach to me personally.

Author: How would you get around having this burden? My assumption is that each instance type wouldn't have a static number of resources that match the `smarter-devices/fuse` so you would need some dynamic way to configure the quantities for each of the different instance types, or at least a collection of instance types.

This becomes a problem when extending out to other resources. Even assuming that there is a consistent quantity of resources for the `smarter-devices/fuse` resource across _all_ instance types, that certainly wouldn't be true for other extended resources.

> Also, we would need to keep the list of available instance types up to date
This works as an override mechanism, so you would still have those instance types available if they didn't have a matching CR for them. The only difference there would be that the extended resource wouldn't be mapped onto it, which would still be something that would need to be addressed with the other "collection-based" approach.

Author: Also, the other thing that we were thinking with the approach proposed in this design, is we would offer some tooling/guidance around how to programmatically generate these values into CRs, so that the maintenance burden is reduced.

For an instance type that has no InstanceType overrides, we would use the default values used in the `createOfferings` call that is generated from the `DescribeInstanceType` call. This set of offerings would describe the full combination of options that are available from the cloud provider for a given instance type. When we schedule and attempt to launch an instance, we validate that there is an offering that matches the nodeTemplate requirements that are added from the scheduler.

In the case that the user provides an Instance Type offerings overrides (currently the scenario that they would do this would be for pricing overrides), we will use the provided offerings from the `InstanceType` CR as the offering values as opposed to the offering requirement values that are used by default for the cloud provider. To simplify the semantic, we assume the “In” operator for the offerings and we only allow a single value for each label requirement.

```yaml
offerings:
  - karpenter.sh/capacity-type: spot
    topology.kubernetes.io/zone: us-west-2b
    karpenter.sh/price: "0.40"
  - karpenter.sh/capacity-type: on-demand # No zonal constraint specified
    karpenter.sh/price: "0.60"
```

> This implies that I can create offerings that don’t contain all constraints that are imposed by the default offering implementation. For instance, in the above `offerings` overrides, I don’t need a zonal constraint on the second offering constraint set which is essentially saying “give me an on-demand offering in any zone and all on-demand offerings regardless of zone cost ~$6.00 per day.”

Reviewer: Instead of "I", perhaps use "you" or "someone" in this paragraph.

1. Number of pods running on a given node
2. The size of the images running on a given node

Thus, we can break down resource overhead into two main categories:

1. Flat resource overhead
    1. This includes systemReserved values for system daemons, OS overhead, etc.
    2. Hard eviction threshold for the node
    3. OS memory overhead (assigned to **VM_MEMORY_OVERHEAD** now) to run the instance
2. Dynamic Resource Overhead (per-pod overhead)
    1. This includes kubeReserved values needed for the container runtime. These values are typically dependent on the number of pods that are scheduled to the instance (or at worst the maximum number of pods that could be scheduled to the instance)
    2. This includes needed ephemeral storage since large images that come from multiple pods scheduled to a node can increase the needed ephemeral storage and affect the need to have larger blockDeviceMappings on the provisioner

Based on these two categories, we can tackle user-defined overhead/resource values in the following way:

1. Flat Resource Overhead

>     1. `kubeletConfiguration` can live at the `Provisioner` level. If there is need to specify flat `kubeReserved` or `systemReserved` values based on the user’s knowledge of the OS/AMI, they are able to set this at the Provisioner-level

Reviewer: ```suggestion
    1. `kubeletConfiguration` can live at the `Provisioner` level. If there is need to specify flat `kubeReserved` or `systemReserved` values based on the user’s knowledge of the OS/AMI, they are able to set this at the Provisioner level
```

    2. Hard eviction thresholds can be set within the `kubeletConfiguration` at the `Provisioner` level as this will often be some globally defaulted value or vary based on knowledge of the OS or the system daemons
    3. OS memory overhead (**VM_MEMORY_OVERHEAD**) can be moved into this `InstanceType` CRD. Since this overhead in bytes will vary on a per-instance type level, we can allow the user to set a custom value to reserve for VM_OVERHEAD rather than defaulting to the percentage value that is currently reserved for overhead.
2. Dynamic Resource Overhead (per-pod overhead)
    1. [RuntimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) provides a mechanism for users to specify per-pod overhead values that can be reserved for each pod. This is the recommended way that Kubernetes provides for specifying any pod overhead values that are dynamic and based on the CRI.
       1. 🔑 Dynamic pod overhead that is tied to the container runtime should move to a `RuntimeClass` if there is pod overhead unique to the CRI that is used by your nodes.

>     2. Other per-pod overhead that may be tied to things like image size can be tied to the pod `ephemeral-stroage` requests which will help Karpenter ensure that it accurately provisions pods to nodes while also ensuring that there is parity between the scheduling decisions of Karpenter and the kube-scheduler.

Reviewer: ```suggestion
    2. Other per-pod overhead that may be tied to things like image size can be tied to the pod `ephemeral-storage` requests which will help Karpenter ensure that it accurately provisions pods to nodes while also ensuring that there is parity between the scheduling decisions of Karpenter and the kube-scheduler.
```

Author: Good catch 👍

    3. In the meantime, while most users may not be using either mechanism to achieve system-wide pod overheads, we will still calculate the `kubeReserved` values based on pod density. There are still some open questions based on whether we should open up the ability to specify the strategy type (see below)
3. Max Pods
    1. In general, it appears to be an anti-pattern in the Kubernetes community to strictly fix maxPods values (for kubeletConfiguration) on a per-instance type level. Users should be restricting the number of pods on their instances based on pod requests and size of instances as opposed to strict pod density sizing. If users are concerned with Provisioners creating instances that are too large, there is now the option to specify GT or LT operators for setting limits for how large an instance type that is provisioned can get.

>     2. 🔑 For users that continue to utilize and/or need `--max-pods`, we will surface a `pods` value in the `.spec.allocatable` of the `InstanceType`. This `pods` value will be used for pod density for bin-packing and scheduling. Since there would now be a `max-pods` value surfaced at both the `InstanceType` level and at the `Provisioner` level, we will take the `min(provisioner.maxPods, instanceType.maxPods)` to be the pod density for the instance. 

Reviewer 2: IIUC, this is the only field of the instancetype CRD that is capable of influencing user data. I'm a bit wary of this coupling, how it might limit our design decisions in the future, as well as the security implications. Put differently, InstanceType is currently a readonly type and you're proposing to give it write privileges. It currently isn't capable of modifying the configuration of nodes, which is a special privilege of the Provisioner. I'm comfortable using it for binpacking, since this relationship exists today. 

Here's a less controversial alternative that functions identically w/ how `type InstanceType interface` works today, and doesn't sacrifice any use cases.

1. The InstanceType CRD can expose `pods` as a value in `Allocatable`
2. The `maxpods` value can be hardcoded in the provisioner's kubelet configuration
3. Our binpacker will default to what the `InstanceType` says (which may be overriden by the CRD), but if the provisioner sets a value, that's what we'll use.
4. If the customer wants to compute a dynamic (w.r.t. instancetype) max pods value, they can do so in user data, and they just need to make sure that this function agrees with the values they've overridden in the instancetype CRD. 
5. If you need different dynamic max pods values for the same instance type but different provisioners, stop and rethink your architecture.

Author: > as well as the security implications
I'm not quite sure I get the security implications that you are trying to call out here. Maybe you could describe it further. I see that this changes some read/write assumptions about the `InstanceType` and how it interacts with the current code logic but I'm not quite sure how this has security implications for the project.

> Here's a less controversial alternative that functions
I think the other option to call out is that we can surface `podsPerCore` in the `kubeletConfiguration` in the `Provisioner` which would give users another method to have this be a dynamic value without the overhead of having to write custom `userData`. If we are going to say that `pods` in the `.sepc.capacity` can't affect `--max-pods`, I would rather just punt having `pods` included entirely.

       1. These per-instance type pod density values will be passed into the userData that is used to bootstrap the node. Instance type discovery will be performed in the userData script to determine the max pods value to pass to `bootstrap.sh`.

**🔑 Open Questions**

> 1. There is still [users asks](https://github.com/aws/karpenter/issues/1803) for calculating the kubeReserved resources value based on the GKE calculation of the kubeReserved resources value as opposed to the EKS default which is based on pod density. Should we allow some flag on the provisioner that can determine which way to base the calculation on? i.e. `kubeReservedMemoryStrategy`

Reviewer: ```suggestion
1. There is still [users asks](https://github.com/aws/karpenter/issues/1803) for calculating the kubeReserved resources value based on the GKE calculation of the kubeReserved resources value, as opposed to the EKS default which is based on pod density. Should we allow some flag on the provisioner that can determine which way to base the calculation on (i.e. `kubeReservedMemoryStrategy`)?
```

> 1. There is still [users asks](https://github.com/aws/karpenter/issues/1803) for calculating the kubeReserved resources value based on the GKE calculation of the kubeReserved resources value as opposed to the EKS default which is based on pod density. Should we allow some flag on the provisioner that can determine which way to base the calculation on? i.e. `kubeReservedMemoryStrategy`

Reviewer 2: I'd love for us to do the work and figure out containerd overhead in 2022, and then provide customers with an accurate runtime class.

In general, it seems the asks from users specify the ephemeral-storage as a function of pod density on the node. We could consider adding `blockDeviceMappings` to the `InstanceType` CR; however, this places more maintenance overhead on the user and makes the `InstanceType` CR vendor-specific.

`InstanceType` CR with Block Device Mappings:

```yaml
apiVersion: karpenter.sh/v1alpha1
kind: InstanceType
metadata:
  name: "c5.large"
spec:
  allocatable: # This defines the starting resources that the instance has
    custom-device.com/resource: "2"
    memory: 1Gi
    cpu: 2
  provider:
    blockDeviceMappings:
    - deviceName: /dev/xvda
      ebs:
        volumeSize: 100Gi
        volumeType: gp3
        iops: 10000
        encrypted: true
        kmsKeyID: "arn:aws:kms:us-west-2:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab"
        deleteOnTermination: true
        throughput: 125
        snapshotID: snap-0123456789
```

> **🔑 Recommendation:** Based on the additional overhead and muddiness of putting `blockDeviceMappings` in the `InstanceType` CR, we will punt on block device mappings for now. For now, we will consider extending ephermal-storage through some dynamic mechanism out of scope, though the hope is that we will have some mechanism as a future feature that calculates the combined needed ephemeral-storage based off of the pod requests, kubeReserved and systemReserved.

Reviewer: ```suggestion
**🔑 Recommendation:** Based on the additional overhead and muddiness of putting `blockDeviceMappings` in the `InstanceType` CR, we will punt on block device mappings for now. For now, we will consider extending ephemeral-storage through some dynamic mechanism out of scope, though the hope is that we will have some mechanism as a future feature that calculates the combined needed ephemeral-storage based off of the pod requests, kubeReserved and systemReserved.
```

### Launch Templates

For now, we are avoiding the need for Instance Type changes to affect launch template cardinality. Configuration at the `InstanceType` level is intended to configure the starting point for the scheduler to know how many resources can be used by the bin-packer. The expectation is that when the node comes up this is roughly the allocatable amount of resources the node will have, less kubeReserved, systemReserved, and evictionThresholds. Thus, the calculation we will do will be:

```

> Allocatable = Instance Type Resources - kubeReserved - systemReserved - evictionThreshhold

Reviewer 2: Now that I see this, my recommendation elsewhere to have the field be `InstanceType.Spec.Allocatable` is inaccurate. Is it more correct to make it `InstanceType.Spec.Capacity` and then we substract off these additional 3 values that are controlled at the provisioner level?

Author: Yes, you are completely right. This should be `Capacity` not `Allocatable`

1. Create a `PricingConfiguration` CRD or ConfigMap that contains mappings between offering and the pricing for that offering. Any offerings specified in this `PricingConfiguration` CRD would be read and overriden when generated the pricing data that Karpenter uses for scheduling
    1. Pros
        1. Creates clean separation between pricing configuration for instances and the system/requirements specific information for instances (keep separation between pricing and `InstanceTypeSettings`
    2. Cons
        1. Creates another CR/configMap that the user has to configure, understand, and maintain
2. Add pricing information to the `InstanceType` CRD
    1. Pros
        1. Maintains all instance-specific information in one location
        2. Less CRDs for the user to maintain
    2. Cons

>         1. Pricing information is instance-type specific which means we are necessarily requiring a 1-1 mapping between some CRD and an instance type. This creates less flexibility if a user wants to describe common settings across an instance family or across an architecture type.

Reviewer: ```suggestion
        1. Pricing information is instance-type specific which means we are necessarily requiring a one-to-one mapping between some CRD and an instance type. This creates less flexibility if a user wants to describe common settings across an instance family or across an architecture type.
```

# Instance Type Settings

## Goals

Allow users to:

> * Define [custom device resources](https://github.com/aws/karpenter/pull/2161) on an instance type

Reviewer 4: nit: Can we reference the issue instead of the pull request here? 
```suggestion
* Define [custom device resources](https://github.com/aws/karpenter/issues/1900) on an instance type
```
